XWIKI-20541: Integrate a WCAG 2.1 validation tool in XWiki build/CI 

Sereza7 opened the pull request:
**Jira:**  https://jira.xwiki.org/browse/XWIKI-20541
**PR Changes:**

- Integrated Axe core to the XWiki build and in the CI pipeline
- Removed the old Dutch Web Guidelines tests

Design proposal here: https://design.xwiki.org/xwiki/bin/view/Proposal/WCAG21validationinbuild
Forum topic here: https://forum.xwiki.org/t/wcag-2-1-validation-proposal/11792
Example of an accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameter xwiki.test.ui.wcag set to true:
[wcagWarnings.txt](https://github.com/xwiki/xwiki-platform/files/10795333/wcagWarnings.txt)

tmortagne commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-docker/pom.xml:
```diff
@@ -156,6 +156,12 @@
       <artifactId>commons-httpclient</artifactId>
     </dependency>
+    <!-- Used by the wcag validation-->
+    <dependency>
```
This dependency is already coming with xwiki-platform-test-ui AFAICS.

tmortagne commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -26,24 +26,31 @@
  * Represents the common actions possible on all Pages.
  * 
- * @version $Id$
+Added  * @version $Id$
```
Looks like a mistake.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/pom.xml:
```diff
@@ -64,13 +64,19 @@
       <groupId>io.github.bonigarcia</groupId>
       <artifactId>webdrivermanager</artifactId>
-      <version>5.1.1</version>
+      <version>5.3.1</version>
```
was this needed or is it something separate?

Sereza7 replied:
This dependency is used in the `com.deque.html.axe-core` package too, and leaving the version to 5.1.1 fails the build.

The build logs without this version change:

>  [INFO] --- maven-enforcer-plugin:1.4.1:enforce (enforce-upper-bounds) @ xwiki-platform-test-ui ---
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for io.github.bonigarcia:webdrivermanager:5.1.1 paths to dependency are:
+-org.xwiki.platform:xwiki-platform-test-ui:15.1-SNAPSHOT
  +-io.github.bonigarcia:webdrivermanager:5.1.1
and
+-org.xwiki.platform:xwiki-platform-test-ui:15.1-SNAPSHOT
  +-com.deque.html.axe-core:selenium:4.6.0
    +-io.github.bonigarcia:webdrivermanager:5.3.1
]

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -118,6 +125,38 @@ public BasePage()
+        }
+        else{
+            LOGGER.info("[{} : {}] This combination of URL:class was already WCAG checked.", this.getPageURL(), this.getClass());
+        }
```
should be put in a separate method to be more clear about what it's about.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -118,6 +125,38 @@ public BasePage()
         super();
         waitUntilPageIsReady();
+
+        Logger LOGGER = LoggerFactory.getLogger(BasePage.class);
```
only needed if WCAG validation is enabled.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -118,6 +125,38 @@ public BasePage()
+            long startTime = System.currentTimeMillis();
+            XWikiWebDriver driver = this.getDriver();
+            try {
+                if (driver == null) {
```
can this happen?

Sereza7 replied:
I shouldn't happen, since the getDriver returns a driver that's always initialized during construction of PersistentTestContext. The instantiation of the driver happens either in the PersistentTestContext constructor or beforehand.
The driver being null would surely make the test framework fail before reaching this point anyways.

I'm removing this clause for the next commit :+1:

vmassol commented:
Thanks @Sereza7 . Does it work? :) Could you show some results of the build when executing it?

Sereza7 commented:
> Thanks @Sereza7 . Does it work? :) Could you show some results of the build when executing it?

Example of an accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameter xwiki.test.ui.wcag set to true:

`mvn clean verify -Dgradle.cache.local.enabled=false -Dgradle.cache.remote.enabled=false -Pintegration-tests -Dxwiki.test.ui.wcag=true > wcagAccessibilityLogs.log`


[wcagAccessibilityLogs.log](https://github.com/xwiki/xwiki-platform/files/10801862/wcagAccessibilityLogs.log)
[wcagWarnings.txt](https://github.com/xwiki/xwiki-platform/files/10795333/wcagWarnings.txt)

vmassol commented:
> Example of an accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameter xwiki.test.ui.wcag set to true:

Sounds good.

I think we should log in the console only when `xwiki.test.debug` is `true` for:
* `10:06:09.981 [main] INFO  org.xwiki.test.ui.po.BasePage - [http://host.testcontainers.internal:8080/xwiki/bin/view/Applications/WebHome : class org.xwiki.test.ui.po.ViewPage] Found [2] WCAG violations.``
* `10:06:22.061 [main] INFO  org.xwiki.test.ui.po.BasePage - [http://host.testcontainers.internal:8080/xwiki/bin/view/Applications/WebHome : class org.xwiki.test.ui.po.ViewPage] This combination of URL:class was already WCAG checked.`

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+  private static final List<String> VALIDATE_TAGS = Arrays.asList("wcag2a", "wcag2aa", "wcag21a", "wcag21aa");
+  private static final List<String> DISABLED_RULES = Arrays.asList();
+
+  private static class WCAGTestResults {
```
A JavaDoc comment that explains the purpose of this class would be nice.

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+    public String getWarnReport() {
+      return this.warnReport;
+    }
+    protected boolean failIsEmpty() {
```
I think it is a good practice to have the verb first, i.e., `isFailsEmpty` (but not sure, this is really just a feeling).

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+      return this.warnReport;
+    }
+    protected boolean failIsEmpty() {
+      return (this.failReport.equals(""));
```
The `()` around the return value aren't necessary and should be omitted as there is no good reason to have them.

michitux reviewed (changes requested):
From the many formatting problems I spotted here I assume that you're either not using an IDE or you haven't properly configured your IDE to follow XWiki code style. We're currently not enforcing code style using CheckStyle on test code (which is a mistake imho) but we should still follow the rules as much as possible so please make sure you're using a properly configured IDE.

Overall, this looks okay, but for me it is not clear how the enabled (or disabled?) rules were chosen. 

Also, from what I understood, this doesn't seem to fail any test so we won't notice the failures unless we look at the results. With this approach, it seems hard to follow the progress we make and to catch regressions. What's the plan how we see the results? Will they be available on CI?

Sereza7 commented on xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/AllTests.java:
```diff
@@ -53,9 +53,13 @@ public static Test suite() throws Exception
         CustomDutchWebGuidelinesValidator dwgValidator = new CustomDutchWebGuidelinesValidator();
         addTest(suite, CustomDutchWebGuidelinesValidationTest.suite(CustomDutchWebGuidelinesValidationTest.class,
             dwgValidator), DefaultValidationTest.class);
+        */
```
This is the last part of this issue. 
https://jira.xwiki.org/browse/XWIKI-20541
The Dutch Web Guidelines Validator was used to check for a set of outdated accessibility standards based on WCAG 2.0 .
I'm removing the concerned files appropriately in the upcoming commit.

Sereza7 replied:
Okay, I'll leave them for now and try to focus making sure the properties they cover are made failing in the new test asap. This should be easy but some rules might not match exactly and catch violations that didn't get caught before, so it's not just a matter of switching the FAILS_ON_RULE entries to true.

michitux replied:
I'm not saying that the new tests need to be equivalent, it's just that we actually have failures of these validation tests from time to time (e.g., they found an input without a label recently) and I think we shouldn't loose this basic level of checks as long as the framework isn't actually active in a way that we notice (i.e., that lead to actual failures that are visible in the CI).

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -1427,6 +1432,46 @@ public String getSecretToken()
+        if (wcagContext.isWcagEnabled()) {
+            float totalTime = (float) wcagContext.wcagTimer / 1000;
+            logger.info("Time spent on WCAG validation [{}] (in s)", totalTime);
+            File wcagDir = new File("target/wcag");
```
For next commit, I set this value as a static final variable in TestUtils (and built it using `File.separator` instead of just a `/` ).
This should be better but I'm not sure this solves this issue completely, I'm not sure of what the best practice here would be.

michitux replied:
I've just had a look and found https://github.com/xwiki/xwiki-platform/blob/e44c811a38d395a58722110e914ee8a1e6dfb00c/xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-docker/src/main/java/org/xwiki/test/docker/internal/junit5/DockerTestUtils.java#L273 and https://github.com/xwiki/xwiki-platform/blob/a25ac4c3dfacd781d6ed1a86d40ca517a7da26e6/xwiki-platform-core/pom.xml#L340 used at https://github.com/xwiki/xwiki-platform/blob/8793d8438ebfb7cfe40d9b4ce31d9f791f563187/xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestDebugger.java#L63 - I think you could/should do something similar for the WCAG results.

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+   * There is a parameter for rules *reviewOnFail* in axe-core, however it's not possible to change it using the
+   * current selenium integration library.
+   **/
+  private static final Map<String, Boolean> FAILS_ON_RULE = Map.<String, Boolean>ofEntries(
```
This was highlighted by Intellij as a code quality warning:
`Vararg method call with 50+ poly arguments may cause compilation and analysis slowdown `

I'm omitting it for the next commit :+1:

michitux replied:
Oh, okay, I didn't know that one, I'm normally just seeing the warning the other way around. You can also keep it and just add a comment that explains why it is there.

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+
+public class WCAGContext{
+
+  /** List of rules to fail on.
```
I needed to update this doc, I'm adding precisions to try and make those points clearer :ok_hand:

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WCAGContext.java:
```diff
@@ -0,0 +1,295 @@
+   * Checks if there's a need to perform an accessibility analysis of a basePage.
+   *
+   */
+  public boolean checkAccessibility(String url, Class pageClass) {
```
I renamed it `isAlreadyInCache` and changed a bit it's behavior, thanks for the tip, it was a really bad name.

vmassol replied:
Note that `isCached()` is also simple and effective ;)

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -118,6 +125,38 @@ public BasePage()
+            XWikiWebDriver driver = this.getDriver();
+            try {
+                if (driver == null) {
+                    throw new NullPointerException("The provided webDriver is null.");
```
https://github.com/xwiki/xwiki-platform/pull/2092#discussion_r1114285333
I removed this throw :+1:

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -1427,6 +1432,46 @@ public String getSecretToken()
+                WCAGContext.writeWCAGReportToFile(failsFile, wcagContext.buildFailsReport());
+            }
+            // Assert the results of the different WCAG Checks are all empty
+            assertFalse(wcagContext.hasWCAGFails(), "There are some accessibility issues in the test suite." +
+              "See target/wcag/wcagFails for more details.");
```
To me it seems unexpected that the code to write the results contains the assertion that fails the build. Also, wouldn't it be nicer to include at least the description of the first failure in the assertion such that it is displayed in Jenkins and whoever sees this doesn't always need to look at the output files?

Sereza7 commented:
**PR Changes:**
* Fixed checkstyle requirements.
* New class XWikiCustomAxeReporter to contain most of the string building for the report, and allow generating a report more adapted to our use case.
* Changed log info to debug.
* Removed hardcoded file location.
* Added back the DutchWebGuidelines (to keep until some fails are triggerable)
* Added the first rule failed details to the error message when asserting there's no failing violation.

Example of an accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameter xwiki.test.ui.wcag set to true. For the sake of variety, one rule has been set as failing:
WcagContext.FAILS_ON_RULE -> entry("list", true)

mvn clean verify -Dgradle.cache.local.enabled=false -Dgradle.cache.remote.enabled=false -Pintegration-tests -Dxwiki.test.ui.wcag=true > wcagAccessibilityLogs.log

[wcagAccessibilityLogs.log](https://github.com/xwiki/xwiki-platform/files/10815444/wcagAccessibilityLogs.log)
[wcagFails.txt](https://github.com/xwiki/xwiki-platform/files/10815447/wcagFails.txt)
[wcagWarnings.txt](https://github.com/xwiki/xwiki-platform/files/10815448/wcagWarnings.txt)

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/XWikiCustomAxeReporter.java:
```diff
@@ -0,0 +1,120 @@
+     * @param scanResults    the list of rules breached during validation.
+     * @return true if there is any rule breached in scanResults
+     */
+    static boolean getReadableAxeResults(String testMethodName, String pageClassName,
```
From what I can see, the result is never used when returning `false` so I wonder why this method doesn't simply return a `String` (or `Optional<String>`) instead of this strange design with a static result string.

Sereza7 replied:
It can return false, if the report asked for is "empty".
I however still set up conditions against empty reports so I'll remove this structure and make sure it still runs as expected.

michitux replied:
Yes, I understand that, but in this case we don't use the string it built so instead of returning a boolean, it could just return the string and when the report is "empty" it could return either an empty string, `null` or an empty `Optional<String>` if you want to be more explicit. That way, it wouldn't need this really ugly structure with the static variables (which are completely incompatible with multi-threading for example - I know this is not relevant here, but still, it's just not nice).

michitux commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,413 @@
+                this.failReport = XWikiCustomAxeReporter.getAxeResultString();
+                if (shortFailReport.isEmpty())
+                {
+                    /* Only store the first failed rule to display in the error logs. */
```
I know this was my suggestion, we could imho also just display all failed rules if this requires extra work to get the first result. My intention was primarily to display something informative when a test fails such that the developer doesn't first need to get the log to see the reason for the failure. My understanding is that we will only enable rules for which we don't have known failures so in this case this also shouldn't flood the logs unless a regression is introduced (or a new test is added, but then it should be fixed immediately).

Sereza7 replied:
The thing is, for less memory use, I process reports as soon as they are generated so I only keep the string we want. This is less flexible than keeping everything in memory and waiting for the end to generate reports.
This makes it a bit weird to save only the first rule break, since it's often only a part of one validation check.
It'd be easier to display the first report fully, however it could get quite lengthy if a few rules fail at once.
This shouldn't happen too often, so I'll remove this static variable and go with the easy solution

michitux replied:
Sounds good, we can change it also later if we should discover that it is indeed flooding the log.

michitux reviewed (changes requested):
Some new comments, but this looks already much better.

Sereza7 commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/XWikiCustomAxeReporter.java:
```diff
@@ -0,0 +1,120 @@
+ */
+/*
+ * Alternatively, at your choice, the contents of this file may be used under the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
```
Should I just copy the copyright notice from here and add a link to this file?
https://github.com/dequelabs/axe-core-maven-html/blob/develop/selenium/src/main/java/com/deque/html/axecore/selenium/AxeReporter.java

michitux replied:
Yes, I think you should copy the copyright notice and include a reference to the source. I think in the end, in particular with the refactoring I suggested, the similarity to the original won't be that big and I think there is not much creativity and thus not much to copyright in this code. But still, it is good to properly attribute the source.

Sereza7 commented:
**PR Changes:**
* Removed useless complexity from XWikiCustomAxeReporter
* Fixed licencing on XWikiCustomAxeReporter
* Changed the error message to display more (simpler/cleaner implementation)
* Fixed @since on files
* Fixed logging for basePage, now following best practice with a static logger.

**Example:**
Accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameter xwiki.test.ui.wcag set to true. For the sake of variety, one rule has been set as failing:
WcagContext.FAILS_ON_RULE -> entry("list", true)

mvn clean verify -Dgradle.cache.local.enabled=false -Dgradle.cache.remote.enabled=false -Pintegration-tests -Dxwiki.test.ui.wcag=true > wcagAccessibilityLogs.log

[wcagAccessibilityLogs.log](https://github.com/xwiki/xwiki-platform/files/10822793/wcagAccessibilityLogs.log)

[wcagFails.txt](https://github.com/xwiki/xwiki-platform/files/10822796/wcagFails.txt)
[wcagWarnings.txt](https://github.com/xwiki/xwiki-platform/files/10822797/wcagWarnings.txt)

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-docker/src/main/java/org/xwiki/test/docker/junit5/TestConfiguration.java:
```diff
@@ -484,6 +492,22 @@ public void setVNC(boolean vnc)
+    /**
+     * @return true if WCAG rules should be checked.
+     */
+    public boolean isWcagEnabled()
```
Missing `@since` annotation

vmassol replied:
Small preference to use WCAG and not `Wcag` since it's an acronym. So either `isWCAG()` or `isWCAGEnabled()`. See also `getExtraJARs()`, `getJDBCDriverVersion()`, etc.

vmassol replied:
note that `isWCAG()` might be more consistent with the other `isVerbose()`, `isOffice()`, etc.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -214,6 +219,8 @@
     /** Cached secret token. TODO cache for each user. */
     private String secretToken = null;
 
+    /** Cached accessibility results. */
```
We use multiline format in general.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -1427,6 +1434,73 @@ public String getSecretToken()
+     *
+     * @return WCAG Context
+     */
+    public WcagContext getWcagContext()
```
`WCAGContext`

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -1427,6 +1434,73 @@ public String getSecretToken()
+     * Writes the wcag validation results to custom reports.
+     * @throws IOException can be thrown when the directory creation fails.
+     */
+
```
extra new line

vmassol replied:
missing `@since`

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/TestUtils.java:
```diff
@@ -1427,6 +1434,73 @@ public String getSecretToken()
+        }
+        return testClassesDirectory;
+    }
+
```
I'd move everything related to WCAG inside another class.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+    /**
+     * @return the first non-empty fail report from all the validations.
+     */
+    public String getFirstWcagFail()
```
`getFirstWCAGFailure()`? The javadoc doesn't say what it does if there's no failure. Not sure that returning an empty string is great. I'd prefer null.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+        for (WCAGTestResults wcagResult : wcagResults)
+        {
+            if (!wcagResult.isFailEmpty())
+            {
```
wrong code style

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+
+    /**
+     * Sets the current test class name.
+     * @param testClassName new value to be set.
```
missing `the`. See code style at https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/#HJavadocBestPractices

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+    {
+        this.testClassName = testClassName;
+    }
+    protected String getTestClassName()
```
new line missing.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+                                     new FileOutputStream(outputFile), StandardCharsets.UTF_8))) {
+            writer.write(output.toString());
+        } catch (IOException e) {
+            e.printStackTrace();
```
This looks wrong

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+        try (Writer writer =
+                     new BufferedWriter(
+                             new OutputStreamWriter(
+                                     new FileOutputStream(outputFile), StandardCharsets.UTF_8))) {
```
bad code style

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/WcagContext.java:
```diff
@@ -0,0 +1,412 @@
+        for (WCAGTestResults result : this.wcagResults) {
+            if (!result.isWarnEmpty()) {
+                if (warningsEmpty) {
+                    mergedReport.append(String.format("WCAG warnings in the test class %s:", getTestClassName()));
```
Not using our logging best practice of using `[...]`. See https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/#HLoggingBestPractices

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/XWikiCustomAxeReporter.java:
```diff
@@ -0,0 +1,128 @@
+    https://github.com/dequelabs/axe-core-maven-html/blob/develop/selenium/src/main/java/com/deque/html/axecore/selenium
+    /AxeReporter.java
+
+ */
```
Weird comment. What does it mean?

Sereza7 replied:
This class was heavily inspired from the axe-core native final class, it's a way to quote the source. Now I'm not sure we have that much from the original source anymore though. 
I'm making it more explicit for the next commit :+1:

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/XWikiCustomAxeReporter.java:
```diff
@@ -0,0 +1,128 @@
+public final class XWikiCustomAxeReporter
+{
+    private XWikiCustomAxeReporter()
+    {
```
not passing checkstykle normally. Should have a comment to explain why it's empty.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -118,6 +126,10 @@ public BasePage()
         super();
         waitUntilPageIsReady();
+
+        WcagContext wcagContext = getUtil().getWcagContext();
```
missing comment and should have a check, there'esno need to gather wcag context if wcag is not enabled

Sereza7 replied:
On this basePage, from what I know, we need to check wcagContext to see if wcag is enabled.

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -693,6 +705,60 @@ private void useShortcutKeyAndLoads(CharSequence... shortcut)
+            Block wcag validation if it is not enabled, in all cases.
+             */
+            return;
+        }
```
to be moved up there

Sereza7 replied:
https://github.com/xwiki/xwiki-platform/pull/2092#discussion_r1115895145
I agreed with @michitux 's opinion, I think it's more important to include the check in the function for explicit validation calls in the UITests. 
However I did not give a wide enough scope (were set as `protected` instead of `public`).

vmassol commented on xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/BasePage.java:
```diff
@@ -693,6 +705,60 @@ private void useShortcutKeyAndLoads(CharSequence... shortcut)
+            /*
+            Performs accessibility validation on the element when instanced if
+            the page is not in the WCAGContext cache.
+             */
```
should use `//`comment-style

Sereza7 commented:
**PR Changes:**
* Corrected some checkstyle issues.
* Added/ fixed incomplete documentation
* Fixed presentation of the report.
* Moved most of the methods from TestUtils to WCAGUTils.

**Example:**
Accessibility warning report generated while building xwiki-platform-menu-test-docker with the parameterxwiki.test.ui.wcagset to true. For the sake of variety, one rule has been set as failing:
WcagContext.FAILS_ON_RULE -> entry("image-alt", true)

>    mvn clean verify -Dgradle.cache.local.enabled=false -Dgradle.cache.remote.enabled=false -Pintegration-tests -Dxwiki.test.ui.wcag=true > wcagAccessibility.log

[wcagAccessibility.log](https://github.com/xwiki/xwiki-platform/files/10882896/wcagAccessibility.log)

[wcagFails.txt](https://github.com/xwiki/xwiki-platform/files/10882897/wcagFails.txt)
[wcagWarnings.txt](https://github.com/xwiki/xwiki-platform/files/10882898/wcagWarnings.txt)
